tools/shared/src/main/scala/org/scalajs/core/tools/optimizer/OptimizerCore.scala

```diff
@@ -2183,6 +2243,32 @@ private[optimizer] abstract class OptimizerCore(
           case _               => default
         }
 
+      case String_+ =>
+        val lhs1 = foldToStringForString_+(lhs)
+        val rhs1 = foldToStringForString_+(rhs)
+        @inline def stringDefault = BinaryOp(String_+, lhs1, rhs1)
+        (lhs1, rhs1) match {
+          case (StringLiteral(s1), StringLiteral(s2)) => 
+            StringLiteral(s1 + s2)
+          case (l, r @ StringLiteral("")) =>
+            foldBinaryOp(op, r, l)
+          case (StringLiteral(""), r) if r.tpe == StringType =>
+            r
+          case (l, BinaryOp(String_+, rl, rr)) => 
+            foldBinaryOp(String_+, BinaryOp(String_+, l, rl), rr)
+          case (l @ BinaryOp(String_+, ll, lr), StringLiteral(r)) =>
+            foldBinaryOp(String_+, ll, lr)  match {
```

Comment:
This `foldBinaryOp` is redundant. By construction, all trees "smaller" than the arguments have already been folded. So this call is always going to be `identity`.

This also means the cases below can and should be "inlined" in the `case` above. For example, the first one becomes:
```scala
case (BinaryOp(String_+, llp, StringLiteral(lrp)), StringLiteral(r)) =>
  BinaryOp(String_+, llp, StringLiteral(lrp + r))
```
and the two other cases disappear, since they basically have to fall back to `default` anyway.